docs/dev/upstreamMerge.md

```diff
@@ -0,0 +1,147 @@
+3. **Solve merge conflicts.**
+4. Open a PR into `upstream-<SHORT_COMMIT_HASH>`.
+
+#### **Important Considerations During Conflict Resolution**
+
+1. **Common Cheatcodes**: The following files/constructors have been moved to `foundry-cheatcodes-common` crate (from their respective `foundry-cheatcodes` crate) to be used by the `foundry-zksync-core` crate.
+
+| **Entity** | **Original** | **Current** |
+| --- | --- | --- |
+| `expect::{ExpectedCallData, ExpectedCallType, ExpectedCallTracker}`  | `foundry-cheatcodes::test::expect` | `foundry-cheatcodes-common::expect` |
+| `record::{RecordAccess}`  | `foundry-cheatcodes::evm` | `foundry-cheatcodes-common::record` |
+| `mock::{MockCallDataContext, MockCallReturnData}`  | `foundry-cheatcodes::evm::mock` | `foundry-cheatcodes-common::mock` |
+
+2. **Strategy Pattern**: [PR #781](https://github.com/matter-labs/foundry-zksync/pull/781) introduces the **strategy pattern** where `Executor`, `CheatcodeInspector`, and Backend are **generalized into traits** with **both EVM and EraVM implementations**.
+   - There are also several other pull requests which refactor the pattern quite a bit, so **don't use just that PR as reference**.
+   -  We have to be on the **lookout for places where inline logic in upstream is replaced with a strategy trait call in zksync**. When getting a merge conflict in these cases, the inline logic changes must be implemented for the trait implementations.
+
+3. **Dependency Updates**: `alloy-zksync` (https://github.com/popzxc/alloy-zksync) might need to be updated to use the alloy version used by foundry (Sample: https://github.com/popzxc/alloy-zksync/pull/39).
+
+### 4. Compilation Error Fixes
+
+With conflict resolutions merged into `upstream-<SHORT_COMMIT_HASH>` branch, the next step is to solve compilation errors. For that:
+1. **Create a new branch `upstream-<SHORT_COMMIT_HASH>-build**`**.
+2. Make `cargo build` work. 
+3. Open a PR solving all compilation errors.
+
+
+### 5. CI and Test Fixes
+
+Once the compilation errors are solved (the previous PR `upstream-<SHORT_COMMIT_HASH>-build` is merged), **make ad hoc changes to make the CI pass**:
+
+  1. Fix **tests**
+  2. Fix **`clippy` errors**
+  3. Resolve **`cargo deny` dependency issues**
+  4. Add **necessary tests** (preferably as separate PRs).
+
+**Note**: This can technically be another PR to review significant changes. Or multiple PRs per functionality fix. As this is an ad-hoc process, define the scope of the PRs as needed.
+
+### 6. Main Branch Integration
+
+With everything solved and the final PR approved, **squash merge it into `main`**.
+
+#### **Why squash merge?**
+This will include all the upstream commits that we will need to include later, but **for now, it's a better alternative than making tricky rebases to remove "bad commits" that are in an intermediate state due to all the merges that were done during the process** (bringing upstream and also updating the branches with ongoing changes on `main`).
```

Comment:
```suggestion
This will include all the changes from upstream but squashed under a single commit. The distinct upstream commits will be included in a later step. We prefer the squash-merge strategy instead of rebase as rebasing would usually involve fixing the merge conflict multiple times with each included upstream commit, often becoming tedious.
```